pyqtgraph/functions.py

```diff
@@ -293,75 +343,115 @@ def mkColor(*args):
-    
-    color = kargs.get('color', None)
-    width = kargs.get('width', 1)
-    style = kargs.get('style', None)
+    # print('mkPen called:',args,kargs)
+    while ( # unravel single element sublists
+        ( isinstance(args, tuple) or isinstance(args,list) )
+        and len(args) == 1 
+    ): 
+        args = args[0]
+    # now args is either a non-list entity, or a multi-element tuple
+    # short-circuits:
+    if isinstance(args, NamedPen):
+        return args # pass through predefined NamedPen directly
+    elif isinstance(args, QtGui.QPen):
+        return QtGui.QPen(args)  ## return a copy of this pen
+    elif isinstance(args, dict): 
+        return mkPen(**args) # retry with kwargs assigned from dictionary
+    elif args is None:
+        return QtGui.QPen( QtCore.Qt.NoPen ) # explicit None means "no pen"
+    # no short-circuit, continue parsing to construct QPen or NamedPen
+    width = kargs.get('width', 1) # width 1 unless specified otherwise
+    if 'hsv' in kargs: # hsv argument takes precedence
+        qcol = hsvColor( *kargs['hsv'] )
+        qpen = QtGui.QPen(QtGui.QBrush(qcol), width)
+    else:
+        if 'color' in kargs:
+            args = kargs['color'] # 'color' KW-argument overrides unnamed arguments
+        if args is None:
+            return QtGui.QPen( QtCore.Qt.NoPen ) # explicit None means "no pen"
+        if args == () or args == []:
+            qpen = NamedPen( 'gr_fg', width=width ) # default foreground color
+        else:
+            result = parseNamedColorSpecification(args)
+            if result is not None: # make a NamedPen
+                name, alpha = result
+                if name == '':
+                    return QtGui.QPen( QtCore.Qt.NoPen ) # empty string means "no pen"
+                qpen = NamedPen( name, alpha=alpha, width=width )
+            else: # make a QPen
+                qcol = mkColor(args)
+                qpen = QtGui.QPen(QtGui.QBrush(qcol), width)
+    # now apply styles according to kw arguments:
+    style = kargs.get('style', None) 
+    width = kargs.get('width', 1) # collect remaining kargs to define properties
     dash = kargs.get('dash', None)
     cosmetic = kargs.get('cosmetic', True)
-    hsv = kargs.get('hsv', None)
-    
-    if len(args) == 1:
-        arg = args[0]
-        if isinstance(arg, dict):
-            return mkPen(**arg)
-        if isinstance(arg, QtGui.QPen):
-            return QtGui.QPen(arg)  ## return a copy of this pen
-        elif arg is None:
-            style = QtCore.Qt.NoPen
-        else:
-            color = arg
-    if len(args) > 1:
-        color = args
-        
-    if color is None:
-        color = mkColor('l')
-    if hsv is not None:
-        color = hsvColor(*hsv)
-    else:
-        color = mkColor(color)
-        
-    pen = QtGui.QPen(QtGui.QBrush(color), width)
-    pen.setCosmetic(cosmetic)
+    assert qpen is not None        
```

Comment:
probably don't want an `assert` in functional code; as they can be squashed/suppressed when you run python with the `-O` argument (which I would expect a fair chunk of our user-base to actually do).

Probably to something like:

```python
if qpen is None:
    raise RuntimeError("some message")
```

Reply:
`qpen` being None is really not supposed to ever happen here; it means I missed a possible execution path. This is an accidentally surviving debug statement.
But I guess throwing a recognizable error would be more helpful. Replaced by
` if qpen is None:`
`  raise ValueError('Failed to construct QPen from arguments '+str(args)+','+str(kargs) )`